mysql: Recognize when client/server negotiate SSL

awelzel opened the pull request:
This instantiates the SSL analyzer when the client requests SSL
so that Zeek now has a bit more visibility into encrypted MySQL
connections.

The pattern used is the same as in the IMAP, POP or XMPP analyzer.

awelzel commented on src/analyzer/protocol/mysql/mysql-analyzer.pac:
```diff
@@ -22,12 +22,19 @@ refine flow MySQL_Flow += {
+		// If the client requested SSL and didn't provide credentials, switch to SSL
+		if ( ${msg.version} == 10 && ( ${msg.v10_response.cap_flags} & CLIENT_SSL ) && ${msg.v10_response.credentials}->empty() )
+			{
+			connection()->zeek_analyzer()->StartTLS();
```
The MySQL header comes with a size and that's used on the top-level unit as &length.

https://github.com/zeek/zeek/blob/33244ae75f21ba01b0c18b2db74615aca822b314/src/analyzer/protocol/mysql/mysql-protocol.pac#L163-L171

So I thought the scenario you point out can't happen because `$input.length()` would not be 0 if the client actually had appended credentials (and the credentials[] array therefore filled) and only 0 if there weren't actual credentials. But I might be assuming things here?

awelzel replied:
Resolving due to probably working out in practice.

awelzel commented:
After chatting with @rsmmr, his main concern was that a `DeliverStream()` call could contain the MySQL handshake packet indicating switching to TLS *and* the first part of the client side TLS handshake when both were delivered together.

I think that concern applies to other analyzers today, too, but will stall this for now.

0xxon commented:
If this is a concern that also applies to other analyzers (which I totally believe - I don't think I ever paid attention to this when adding StartSSL support) - I am of the opinion that we should just add this.

Also - even if we miss the first handshake packet, we still should get the remaining ones - which is much better than what we have now.

awelzel commented:
> If this is a concern that also applies to other analyzers (which I totally believe - I don't think I ever paid attention to this when adding StartSSL support) - I am of the opinion that we should just add this.

I was rather close to ditching this PR because of the (suspected?) BinPAC trickery to pull this off properly, or try another round of "but it's not worse than the others and it's a nice feature" - so thanks for doing the latter, maybe we can push it through?

> Also - even if we miss the first handshake packet, we still should get the remaining ones - which is much better than what we have now.
 
Today we actually cause a violation because there isn't any data left to parse when we expected some more, so seems we're trading this likely analyzer violation with a less likely one when there's too much data?

```1540845683.984788       violation       protocol        MYSQL   Cy7b89xFkY3wL9tpc       -       127.0.0.1       59272   127.0.0.1       3306    Binpac exception: binpac exception: string mismatch at /home/awelzel/corelight-oss/zeek/src/analyzer/protocol/mysql/mysql-protocol.pac:239: \x0aexpected pattern: "[^\\0]*\\0"\x0aactual data: ""    -
```

As a reference, this topic has provoked zeek/spicy#1328 so hopefully the future is brighter/simpler.

awelzel commented:
@0xxon - I've rebased and changed the approach so it's the same as with IMAP, POP3 or IMAP. Would you be up for an approval if you still think we should do it?
